src/openassetio-core/src/utils/path.cpp

```diff
@@ -55,1384 +50,7 @@ namespace utils {
 //  * Should we support literal `/` in verbatim `\\?\` paths - we can
 //    %-encode it.
 
-namespace {
-
-// Non-`file://` URLs are invalid.
-constexpr StrView kErrorNotAFileUrl = "Not a file URL";
-// Path provided to pathToUrl is empty.
-constexpr StrView kErrorEmptyPath = "Path is empty";
-// E.g. empty path or no drive letter in Windows paths.
-constexpr StrView kErrorInvalidPath = "Path is ill-formed";
-// E.g. no leading / in path.
-constexpr StrView kErrorRelativePath = "Path is relative";
-// There's a `..` segment in the path.
-constexpr StrView kErrorUpwardsTraversal = "Path contains upwards traversal";
-// A `\0` was found in the (decoded) path.
-constexpr StrView kErrorNullByte = "Path contains NULL bytes";
-// Decoding a percent-encoded URL reveals an extra path separator.
-constexpr StrView kErrorEncodedSeparator = "Percent-encoded path separator";
-// E.g. Non-ASCII hostname in URL.
-constexpr StrView kErrorUnsupportedHostname = "Unsupported hostname";
-// E.g. Non-ASCII hostname in path.
-constexpr StrView kErrorInvalidHostname = "Path references an invalid hostname";
-// E.g. Windows device path with forward slashes - technically allowed
-// (as a literal rather than path separator) but unsupported by us.
-constexpr StrView kErrorUnsupportedDevicePath = "Unsupported Win32 device path";
-// E.g. `file://server/path` on posix.
-constexpr StrView kErrorNonLocal = "Unsupported non-local file";
-// Ada flagged an error setting the path component of a URL.
-constexpr StrView kErrorInvalidUrlPath = "Invalid URL path";
-// Ada failed to parse a given URL.
-constexpr StrView kErrorUrlParseFailure = "Invalid URL";
-
-// Constants for common character (sets) used in string processing.
-// Useful for greppability.
-constexpr char kColon = ':';
-constexpr char kPercent = '%';
-constexpr char kHyphen = '-';
-constexpr StrView kAnySlash = "\\/";
-constexpr char kForwardSlash = '/';
-constexpr char kBackSlash = '\\';
-constexpr StrView kBackSlashStr = "\\";
-constexpr StrView kDoubleBackSlash = R"(\\)";
-
-/**
- * Throw an exception formatted to contain the problematic string.
- *
- * @param message Error message.
- * @param pathOrUrl Problematic string to append to message.
- */
-void throwError(const StrView message, const StrView pathOrUrl) {
-  throw errors::InputValidationException(fmt::format("{} ('{}')", message, pathOrUrl));
-}
-
-/**
- * Various utility components for handling different types of path/URL
- */
-namespace handlers {
-
-/**
- * Utility for dealing with `/`-separated strings.
- *
- * I.e. posix paths and URLs.
- */
-struct ForwardSlashSeparatedString {
```

Comment:
The `struct`s are for a few reasons.  

* Perhaps the main reason: as you note, the regex(s) mean data storage is required somewhere.
* Single responsibility and component reuse.

Getting a bit more philosophical, although not polymorphic, these structs satisfy a (compile-time) interface.  If templated (and especially if we had C++20 concepts), alternative implementations (e.g. mocks) could be injected.  Templatising/polymorphism isn't necessary for the current use-case/testing, so I stopped short of that.  But I think the pattern is still sound and promotes extensibility and reuse.  IIRC its a bit like the Strategy pattern, if you need a name for it.

Global `static`s of regexes are not thread-safe, crash the entire program if they fail to compile, and are non-trivial types that take time to construct (albeit small, but bad practice, and adds up over lots of libraries that are similarly badly behaved).

Combining all the regexes into a single object was something I considered, but I thought keeping them close to the function where they're used helped discoverability.

Interestingly enough, my first pass was a bunch of free functions with global `static` `Regex` objects, and it was a bit of a mess (albeit largely because it was all in one file), and needed extra state and branching.

I didn't really finish the refactor into these parser components.  I had wondered about combining some, and splitting some others (especially `DrivePath`, which has two responsibilities).

Reply:
Capturing offline discussions: we've gone back and forth on this, to the point of refactoring the whole thing to use free functions.  

This had the disadvantage of passing around an object with all the precompiled regexes, which bloats the functions and loses the encapsulation of the regex and the function where it is used.  It also didn't really help wall-of-text readability.  

It did help to simplify the top-level entry point (no construction of all these utility classes), and had the advantage of allowing the regex namespacing to vary from the function namespacing.

We continue to disagree, but I've reverted back to using the component-based approach, with some tidying at the top-level to hopefully make it a bit easier to digest.

Reply:
Initial tidying as described above 3def16df. Also see splitting into multiple files to hopefully address https://github.com/OpenAssetIO/OpenAssetIO/pull/1246#discussion_r1461030086 and then `const`ing all the things in be8e6168 (I'll make a separate comment about this last one)

Reply:
> We continue to disagree,

Disagreeing agreeably! Resolving under that auspice.